### ISSUE TITLE
some devices , the logcat is really fast , and it often cause the page to die 

### ISSUE THREAD

**Issue by channingbreeze** [openstf/stf#59]
some devices , the logcat is really fast , and it often cause the page to die , how to solve this problem

**Comment by sorccu** [comment]
Do you mean the log function in the STF UI, or just `adb logcat`?

**Comment by channingbreeze** [comment]
the log in stf ui
I find the reason , it's because the logcat puts chinese encoded with utf8 , and the logcat module cannot parse chinese , but i do not know how to make it compatible with chinese .
do you have any suggestions?

**Comment by lukzeg** [comment]
Hi @channingbreeze,

Can you make a simple test. Can you change:
stf/res/app/components/stf/logcat-table/logcat-table-directive.js
ln:16
var maxEntriesBuffer = 3000
to some smaller value, let's say 500 and confirm, that this issue is still visible?
As I think that the problem might be connected with the number of entries which STF is storing in memory?
We need to remember that each of entries stores:
   date <double>
   dateLabel <string>
   deviceLabel <string>
   message <string>
   pid: <int>
   priority <int>
   priorityLabel <string> 
   serial <string>
   tag <string>
   tid <int>
So you can imaginate, that in each of received log, a specially message, memory usage it will increase.
For example in my case nexus 5 which often throws errors on tested app, when I reached 131 entries, the strings which are stored in memory consumes 35mb(in Firefox). So even 500 entries might be a lot, but everything depends mostly from data stored inside message variable.

**Comment by l674964417** [comment]
I also encountered this problem , if I open the logs, sometime later, the browser will die! 
Do you have any solutions about this problem?

**Comment by thinkhy** [comment]
@l674964417 @lukzeg It's indeed a problem that the web page of STF would hang when logcat output at a fast and furious rate. My solution is adding rate limiter both in websocket and device component:

```javascript
    var RateLimiter = require('limiter').RateLimiter
    var logcatRateLimiter = new RateLimiter(10, 'second') // allow 10 msg to be sent every 1 second
    ... ...
    .on(wire.DeviceLogcatEntryMessage, function (channel, message) {
        if (logcatRateLimiter != null) {
             logcatRateLimiter.removeTokens(1, function () {
               if (logcatRateLimiter != null) {
                 socket.emit('logcat.entry', message)
               }
             })
       }
```

@sorccu It might be a performance problem in browse, any suggestion to resolve this kind of problem?

**Comment by qin-young** [comment]
@thinkhy  I use rate limiter, but it cause  that  the log in web is not real-time

**Comment by thinkhy** [comment]
@qin-young yes, it's as expected and better than hitting browser crash ... Myabe the best way is to filter log entries
